### ISSUE TITLE
new rule: param-names

### ISSUE THREAD

**Issue by xjamundx** [eslint/eslint#1626]
I wrote this rule for my team at PayPal. I'm happy to package it up and give it back into the main repo.

`param-names`

Choose a single correct name for common parameters.

Here's an example of how the config would work:
```json
{
    "event": ["e", "evt"],
    "error": ["err"],
    "callback": ["cb"]
}
```

This will will warn is the it sees a param named `cb` and ask you to rename it to `callback`.

The following would all **warn**:

```javascript
function processEvent(evt) {}
function handleError(err){}
function doSomething(cb) {}
```

By default it will be off and provide no guidance.

**Comment by Delapouite** [comment]
Thanks. Do you also use this rule to enforce request / response instead of req / res in express apps?

**Comment by xjamundx** [comment]
@Delapouite yes this is totally possible!

**Comment by Nate-Wilkins** [comment]
@xjamundx Is there a reason this is only param names? Just out of curiosity could be all variable names

**Comment by xjamundx** [comment]
It would be very easy to extend this rule to work for all variable names (or to have a separate rule for such), but IMO param names represent the API (either internal or external) for your application whereas local variables are essentially implementation details that may not benefit as many people from strict consistency. For example if you said you must use `i` instead of `count` that would work, except what if `count` wasn't a counter. At this point you may need to integrate the rule with some sort of type-system for it to become useful :)

**Comment by xjamundx** [comment]
Either way, let me submit the pull request. It sounds like there's some interest.

**Comment by nzakas** [comment]
I need some more convincing that this would be a good addition to the core. We created plugins so we wouldn't have to keep adding new rules ourselves, and I'm a bit torn on whether our not this one should be included.

Also, some questions about how this works: what if `e` should be replaced with either `event` or `error`? Is that supported?

**Comment by Nate-Wilkins** [comment]
@nzakas I don't think this should be included this feels like more of a stylistic rule than anything else.

**Comment by xjamundx** [comment]
Currently it throws if you try to assign the same bad params to different good params (i.e. `e` is a bad version of `event` or `error`). I was planning to change that for this PR to just warn with something like `Please use "error" or "event" instead of "e"`. 

As far "stylistic rules" those can be pretty important too, especially for a large team/project :)

If you guys don't want to include it as an official rule, how about setting up a wiki page where we can post links to custom rules?

**Comment by gcochard** [comment]
You can get a list of available plugins by running `npm search eslint-plugin-`

**Comment by nzakas** [comment]
@Nate-Wilkins we have stylistic rules, that's not a big deal. It's more about being important enough that it's a "must-have" or not.

@xjamundx I'd rather setup a page on eslint.org for that purpose and let people send pull requests to update. That said, the idea of having a spot for authors to add their plugins is a great idea.

**Comment by xjamundx** [comment]
One interesting point here is that I recently enabled this and found that the other rule I had about handling errors was not catching errors if they weren't `err`. When I enabled this rule I found an `error` that was slipping through unhandled. Might be obscure though, but I like it :)

**Comment by nzakas** [comment]
Rather than letting this languish, I'd suggest publishing as a plugin. Part of the rain we have plugins is so rules don't have to be accepted into ESLint core for people to use them.
